### PR TITLE
fix: Add error reporting for Retrieval

### PR DESCRIPTION
**Pull request by jordiclive**
Adding Error reporting as discussed in #867.
(- Add cmd line kwargs) cmd line is under overhaul, will update if gets added to all tasks.

- Separate tasks to be done in another PR

### CODE REVIEW

**Review by isaac-chung — COMMENTED** [review]
Great start! I think we can keep this PR's scope within retrieval first, then extend it to other tasks in a separate PR.

Could you see if you can get it running via command line, e.g. `mteb -m intfloat/multilingual-e5-base -t kardosdrur/estonian-qa --export_errors` ? That way: 
a) we know that it works, and 
b) the kwargs would be properly passed down. Feel free to start from `mteb/cmd.py`

File: mteb/abstasks/AbsTaskRetrieval.py [hunk 1]
```diff
@@ -325,6 +329,39 @@ def _evaluate_subset(
...
+                if false_positives or false_negatives:
+                    errors[qid] = {
+                        "query": queries[qid],
+                        "false_positives": false_positives,
+                        "false_negatives": false_negatives,
+                    }
+
+            errors_save_path = (
+                output_folder / f"{self.metadata_dict['name']}_{hf_subset}_errors.json"
+            )
```
[the remark this answers is not in the document]
Comment (jordiclive) [reply mteb/abstasks/AbsTaskRetrieval.py:361]: Was following the same format save_predictions, as output_folder is pathlib.Path

**Comment by jordiclive** [comment]
@isaac-chung , Great! Happy to work on this and for other tasks.

Does `mteb -m intfloat/multilingual-e5-base -t kardosdrur/estonian-qa` work for you?
```
File "mteb/mteb/overview.py", line 230, in get_task
    return TASKS_REGISTRY[task_name]().filter_languages(languages, script)
           ~~~~~~~~~~~~~~^^^^^^^^^^^
KeyError: 'kardosdrur/estonian-qa'
```

I think kwargs like save_predictions are not supported in the cmd line, would we need a change like #877

File: mteb/abstasks/AbsTaskRetrieval.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply mteb/abstasks/AbsTaskRetrieval.py:361]: I would in general use self.metadata.name (we are trying to move away from the _dict.

**Comment by isaac-chung** [comment]
My suggestion is to add stuff like the command line option in this PR. At the end of the day, we want to be sure that whatever changes made will work. So that could be an added test, or extending the needed components to support running it on the command line.

> Does mteb -m intfloat/multilingual-e5-base -t kardosdrur/estonian-qa work for you?

Oh my bad. Should have used the task name, not the dataset name. Try `mteb -m intfloat/multilingual-e5-base -t EstQA` instead?

**Comment by jordiclive** [comment]
@isaac-chung Great thanks, I added kwargs option (for save_predictions as well) in this PR and checked it works with
`mteb -m intfloat/multilingual-e5-base -t EstQA --export_errors True`

File: mteb/abstasks/AbsTaskRetrieval.py [hunk 2]
```diff
@@ -333,6 +336,39 @@ def _evaluate_subset(
...
+                ]
+                if false_positives or false_negatives:
+                    errors[qid] = {
+                        "query": queries[qid],
+                        "false_positives": false_positives,
+                        "false_negatives": false_negatives,
+                    }
+
+            errors_save_path = (
+                output_folder / f"{self.metadata.name}_{hf_subset}_errors.json"
```
Comment (isaac-chung) [inline mteb/abstasks/AbsTaskRetrieval.py:367]: Results now follow a new structure that requires a model revision (see https://github.com/embeddings-benchmark/mteb/pull/842). That said, I think this is OK for now until we agree on how to pass model revision to this point (e.g. kwargs / via the RetrievalEvaluator class). CC @KennethEnevoldsen

**Review by isaac-chung — APPROVED** [review]
Thanks for iterating! Just a few small comments. I'll also enable the tests now.

File: mteb/cmd.py [hunk 3]
```diff
@@ -126,7 +126,19 @@ def main():
...
+    if additional_kwargs:
+        for i in range(0, len(additional_kwargs), 2):
+            key = additional_kwargs[i].lstrip("-")
+            value = additional_kwargs[i + 1]
+            if value.lower() in [
+                "true",
+                "false",
+            ]:
+                value = value.lower() == "true"
+            kwargs[key] = value
```
[the remark this answers is not in the document]
Comment (jordiclive) [reply mteb/cmd.py:141]: ok sounds good, I was thinking of other kwargs for eval like save_predictions. But I can also add that one explicitly.

**Comment by jordiclive** [comment]
@isaac-chung thanks, changed kwargs.

**Comment by KennethEnevoldsen** [comment]
@jordiclive and @isaac-chung just letting you both know that I made an overhaul to the cli in #882. From my reading of this PR though they should still be compatible (I can just add the arguments into the mteb run parser).

However, when reading through it does seem problematic to add a flag that only works for some of the tasks. Instead, I would add in the non-cli changes, making it possible to create a Python script for running it locally (or alternatively the new CLI construction can be extended in a python script):

```
python my_mteb_cli.py ... # same as mteb, just with the new arguments added to it
```

In the future, we can then fully fledge out the flag such that errors are saved for all task types.

Let me know what you think about this - the intend here is to make the CLI more transparent not to remove flexibility from the user.

**Comment by jordiclive** [comment]
@KennethEnevoldsen ok so remove the cli? Just like save_predictions cannot be enabled through the cmd line and is a kwarg for `MTEB.run`. In a previous commit it was [optional kwargs](https://github.com/embeddings-benchmark/mteb/blob/e462353bff4acf2a61eecbcb83ff12d1101a19ad/mteb/cmd.py) to cmd line

Yes, I was planning to try and implement this feature for the other tasks.

OK cool with your PR, good to neaten up cmd line and documentation—I think I tried the current main this morning and there were some issues with normal running and output folder.

**Comment by KennethEnevoldsen** [comment]
Yea remove it for now and then we can add it back into the CLI, when it is available for all tasks.

If you experience any issues, please submit an issue and we will make sure to take a look at it.

**Comment by jordiclive** [comment]
ok, done.

Will do.

**Comment by isaac-chung** [comment]
> However, when reading through it does seem problematic to add a flag that only works for some of the tasks. 

@KennethEnevoldsen Yes that makes sense. I originally thought the CLI would have been a convenient avenue to test out the  non-cli changes. should have stuck with the test suite 😅 thanks for taking a look!

**Comment by isaac-chung** [comment]
@jordiclive thanks for your contributions! I'll merge this in now.

**Pull request merged**